File: closure/closure.cpp

#include <vector>
#include <iostream>
#include <memory>

typedef std::function<int()> func;

std::vector<func> generate() {
   std::vector<func> result;

   auto counter = std::make_shared<int>(0);

   for (int i = 0; i < 10; ++i) {
      auto increment = [=]() {
         return (*counter)++;
      };
      result.push_back(increment);
   }

   return result;
}

int main() {
   auto seq = generate();
   auto seq2 = seq;

   for (auto& s : seq) {
      std::cout << s() << std::endl;
   }

   for (auto& s : seq2) {
      std::cout << s() << std::endl;
   }

   return 0;
}

File: closure/closure.swift

// Closures are reference types in Swift.
// They automatically capture variables by reference if they are mutated

func generate() -> [() -> Int] {
    var result:[() -> Int] = []

    var counter = 0

    for _ in 1...10 {
        func increment() -> Int {
            return counter++;
        }
        result.append(increment)
    }

    return result;
}

let seq = generate()
let seq2 = seq

for s in seq {
    print(s())
}

for s in seq2 {
    print(s())
}

